src/petals/client/routing/sequence_manager.py

```diff
@@ -0,0 +1,256 @@
+        self._rpc_info = rpc_info
+
+        if sequence_info is None:
+            self.sequence_info = RemoteSequenceInfo.make_empty(block_uids)
+            self.trigger_update()
+        else:
+            self.sequence_info = sequence_info
+            assert block_uids == sequence_info.block_uids
+            self._thread.ready.set()  # no need to await the first dht fetch
+
+        if start:
+            self.run_in_background()
+
+    def run_in_background(self, await_ready: bool = True, timeout: Optional[float] = None) -> None:
+        """
+        Starts the updater thread in a background. if await_ready, this method will wait until sequence manager
+        is ready to process incoming requests or for :timeout: seconds max.
+        """
+        self._thread.start()
+        if await_ready:
+            self._thread.ready.wait(timeout)
+
+    def make_sequence(self, start_index: int = 0, end_index: Optional[int] = None) -> List[RemoteSpanInfo]:
+        """
+        Form a sequence of remote servers that collectively serve all consecutive layers
+
+        :param start_index: optional index of the first module in a sequence, default = the first of block_uids
+        :param end_index: optional index of the last module (non-inclusive), default = after last of block uids
+        """
+        if not self.is_alive():
+            logger.error("Using a sequence manager that is not running: it has either crashed or never started")
+        if not self.ready.is_set():
+            logger.warning("Remote SequenceManager is still searching for routes, waiting for it to become ready")
+            self.ready.wait()
+
+        end_index = end_index if end_index is not None else len(self)
+        span_sequence = []
+        current_index = start_index
+        while current_index < end_index:
+            candidate_spans = self.sequence_info.spans_containing_block[current_index]
+            chosen_span = random.choice(candidate_spans)  # TODO this should be replaced with proper load balancing
+
+            assert chosen_span.start <= current_index < chosen_span.end
+            span_sequence.append(RemoteSpanInfo(start=current_index, end=chosen_span.end, peer_id=chosen_span.peer_id))
+            current_index = chosen_span.end
+
+        return span_sequence
+
+    def __getitem__(self, ix: Union[int, slice]) -> RemoteSequenceManager:
+        """Get a RemoteSequenceManager for a sub-sequence of blocks"""
+        assert isinstance(ix, (int, slice))
+        if not isinstance(ix, slice):
+            ix = slice(int(ix), int(ix) + 1, 1)
+        return type(self)(
+            self.dht,
+            self.block_uids[ix],
+            self.p2p,
+            update_period=self._thread.update_period,
+            timeout=self.timeout,
+            min_backoff=self.min_backoff,
+            sequence_info=self.sequence_info[ix],
+            rpc_info=self._rpc_info,
+            start=True,
+        )
+
+    def trigger_update(self):
```

Comment:
Let's rename:
- `trigger_update()` -> `update()` (maybe with `wait=True/False`)
- `update_()` -> `_update()` (it's a private method, a user can't call it because it's not thread-safe)